packages/tx-construction/test/tx-builder/TxBuilderDelegatePortfolio.test.ts

```diff
@@ -0,0 +1,413 @@
+          name: 'portfolio',
+          pools: [
+            {
+              id: Cardano.PoolIdHex(Cardano.PoolId.toKeyHash(poolIds[1])),
+              weight: 1
+            }
+          ]
+        })
+        .build()
+        .inspect();
+
+      expect(tx.body.certificates?.length).toBe(1);
+      expect(tx.body.certificates![0]).toEqual({
+        __typename: Cardano.CertificateType.StakeKeyDeregistration,
+        stakeKeyHash: Cardano.RewardAccount.toHash(groupedAddresses[0].rewardAccount)
+      });
+    });
+
+    it('portfolio with empty pools array is not a valid CIP17 portfolio', () => {
+      expect(() => txBuilder.delegatePortfolio({ name: 'portfolio', pools: [] })).toThrow();
+    });
+
+    it('derives more stake keys when portfolio has more pools than available keys', async () => {
+      const pools = poolIds.slice(0, 3);
+      const tx = await txBuilder
+        .delegatePortfolio({
+          name: 'portfolio',
+          pools: pools.map((pool) => ({ id: Cardano.PoolIdHex(Cardano.PoolId.toKeyHash(pool)), weight: 1 }))
+        })
+        .build()
+        .inspect();
+
+      const rewardAccounts = await txBuilderProviders.rewardAccounts();
+
+      expect(tx.body.certificates).toEqual([
+        {
+          __typename: Cardano.CertificateType.StakeKeyRegistration,
+          stakeKeyHash: Cardano.RewardAccount.toHash(rewardAccounts[2].address)
+        },
+        {
+          __typename: Cardano.CertificateType.StakeDelegation,
+          poolId: poolIds[2],
+          stakeKeyHash: Cardano.RewardAccount.toHash(rewardAccounts[2].address)
+        }
+      ]);
+    });
+  });
+
+  describe('partial pre-existing multi-delegation', () => {
+    beforeEach(async () => {
+      const txBuilderFactory = await createTxBuilder([
+        { keyStatus: Cardano.StakeKeyStatus.Registered, poolId: poolIds[0] },
+        { keyStatus: Cardano.StakeKeyStatus.Registered }
+      ]);
+      groupedAddresses = txBuilderFactory.groupedAddresses;
+      txBuilder = txBuilderFactory.txBuilder;
+    });
+
+    it('portfolio is superset: adds certificates for the new delegations', async () => {
+      const tx = await txBuilder
+        .delegatePortfolio({
+          name: 'portfolio',
+          pools: [
+            {
+              id: Cardano.PoolIdHex(Cardano.PoolId.toKeyHash(poolIds[0])),
+              weight: 1
+            },
+            {
+              id: Cardano.PoolIdHex(Cardano.PoolId.toKeyHash(poolIds[1])),
+              weight: 1
+            }
+          ]
+        })
+        .build()
+        .inspect();
+
+      expect(tx.body.certificates?.length).toBe(1);
+      expect(tx.body.certificates![0]).toEqual({
+        __typename: Cardano.CertificateType.StakeDelegation,
+        poolId: poolIds[1],
+        stakeKeyHash: Cardano.RewardAccount.toHash(groupedAddresses[1].rewardAccount)
+      });
+    });
+
+    it('changes delegation but does not deregister stake keys that are not delegated', async () => {
```

Comment:
It would deregister the stake key if the stake key was also delegated.
This case is only for a stake key that is already registered, but not delegated. Also it wasn't needed for by the portfolio.
But I agree to your observation and I don't think there is any reason to keep it registered, apart from the fact that we "found" it that way.
I'll make an update to deregister it.

Reply:
Updated here: [4d8c094](https://github.com/input-output-hk/cardano-js-sdk/pull/764/commits/4d8c094b5e96021adaf657ec8a1ed4416e1f985d)